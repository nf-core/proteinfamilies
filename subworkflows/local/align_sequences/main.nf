/*
    MULTIPLE SEQUENCE ALIGNMENT
*/

include { FAMSA_ALIGN } from '../../../modules/nf-core/famsa/align/main'
include { MAFFT_ALIGN } from '../../../modules/nf-core/mafft/align/main'

workflow ALIGN_SEQUENCES {
    take:
    sequences      // tuple val(meta), path(fasta)
    alignment_tool // string: MSA tool

    main:
    ch_versions   = Channel.empty()
    ch_alignments = Channel.empty()

    if (alignment_tool == 'famsa') {
        alignment_res = FAMSA_ALIGN( sequences, [[:],[]], false )
        ch_versions = ch_versions.mix( FAMSA_ALIGN.out.versions )
        ch_alignments = alignment_res.alignment
    } else { // fallback: mafft
        alignment_res = MAFFT_ALIGN( sequences, [[:], []], [[:], []], [[:], []], [[:], []], [[:], []], false )
        ch_versions = ch_versions.mix( MAFFT_ALIGN.out.versions )
        ch_alignments = alignment_res.fas
    }

    emit:
    versions   = ch_versions
    alignments = ch_alignments
}
