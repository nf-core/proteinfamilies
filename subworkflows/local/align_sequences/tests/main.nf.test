nextflow_workflow {

    name "Test Subworkflow ALIGN_SEQUENCES"
    script "../main.nf"
    workflow "ALIGN_SEQUENCES"

    test("fasta - famsa") {

        when {
            workflow {
                """
                alignment_tool = 'famsa'

                input[0] = Channel.of([
                    [ id: 'test_famsa' ], // meta map
                    file(params.pipelines_testdata_base_path + 'proteinfamilies/test_data/subworkflows/chunk_1.fasta', checkIfExists: true)
                ])
                input[1] = alignment_tool
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match() }
            )
        }
    }

    test("fasta - mafft") {

        when {
            workflow {
                """
                alignment_tool = 'mafft'

                input[0] = Channel.of([
                    [ id: 'test_mafft' ], // meta map
                    file(params.pipelines_testdata_base_path + 'proteinfamilies/test_data/subworkflows/chunk_1.fasta', checkIfExists: true)
                ])
                input[1] = alignment_tool
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match() }
            )
        }
    }

    test("fasta - famsa - stub") {

        options "-stub"

        when {
            workflow {
                """
                alignment_tool = 'famsa'

                input[0] = Channel.of([
                    [ id: 'test_famsa_stub' ], // meta map
                    file(params.pipelines_testdata_base_path + 'proteinfamilies/test_data/subworkflows/chunk_1.fasta', checkIfExists: true)
                ])
                input[1] = alignment_tool
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match() }
            )
        }
    }

    test("fasta - mafft - stub") {

        options "-stub"

        when {
            workflow {
                """
                alignment_tool = 'mafft'

                input[0] = Channel.of([
                    [ id: 'test_mafft_stub' ], // meta map
                    file(params.pipelines_testdata_base_path + 'proteinfamilies/test_data/subworkflows/chunk_1.fasta', checkIfExists: true)
                ])
                input[1] = alignment_tool
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match() }
            )
        }
    }

}
