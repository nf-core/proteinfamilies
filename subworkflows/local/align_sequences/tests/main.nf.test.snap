{
    "fasta - famsa - stub": {
        "content": [
            {
                "0": [
                    "versions.yml:md5,cbb230db362dd90a73a8a916b4448d5e"
                ],
                "1": [
                    [
                        {
                            "id": "test_famsa_stub"
                        },
                        "test_famsa_stub_null.aln:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "alignments": [
                    [
                        {
                            "id": "test_famsa_stub"
                        },
                        "test_famsa_stub_null.aln:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,cbb230db362dd90a73a8a916b4448d5e"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-25T10:36:25.613783976"
    },
    "fasta - famsa": {
        "content": [
            {
                "0": [
                    "versions.yml:md5,cbb230db362dd90a73a8a916b4448d5e"
                ],
                "1": [
                    [
                        {
                            "id": "test_famsa"
                        },
                        "test_famsa_null.aln:md5,61de03fefaa5f2501b65ba44f8a3fd2d"
                    ]
                ],
                "alignments": [
                    [
                        {
                            "id": "test_famsa"
                        },
                        "test_famsa_null.aln:md5,61de03fefaa5f2501b65ba44f8a3fd2d"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,cbb230db362dd90a73a8a916b4448d5e"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-25T10:35:29.565556703"
    },
    "fasta - mafft": {
        "content": [
            {
                "0": [
                    "versions.yml:md5,d454b643526f65a83b3f26bfdfeb9f10"
                ],
                "1": [
                    [
                        {
                            "id": "test_mafft"
                        },
                        "test_mafft_null.fas:md5,61de03fefaa5f2501b65ba44f8a3fd2d"
                    ]
                ],
                "alignments": [
                    [
                        {
                            "id": "test_mafft"
                        },
                        "test_mafft_null.fas:md5,61de03fefaa5f2501b65ba44f8a3fd2d"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,d454b643526f65a83b3f26bfdfeb9f10"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-25T10:36:14.655653162"
    },
    "fasta - mafft - stub": {
        "content": [
            {
                "0": [
                    "versions.yml:md5,d454b643526f65a83b3f26bfdfeb9f10"
                ],
                "1": [
                    [
                        {
                            "id": "test_mafft_stub"
                        },
                        "test_mafft_stub_null.fas:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "alignments": [
                    [
                        {
                            "id": "test_mafft_stub"
                        },
                        "test_mafft_stub_null.fas:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,d454b643526f65a83b3f26bfdfeb9f10"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-25T10:36:37.695112353"
    }
}